### PR TITLE
Fix bug in VirtualMachineStuckOnNode alert

### PR DESCRIPTION
**Pull request by sradco**
#### Before this PR:
During VMI phase transitions or live migrations,
kubevirt_vmi_info can have multiple series for the same VM (e.g. phase=scheduling + phase=running,
or two running series on different nodes).
This breaks group_left joins which require right-side uniqueness per match group.

#### After this PR:
Fix VirtualMachineStuckOnNode and VMCannotBeEvicted alerts that were impacted by this.

### References
- Fixes #https://issues.redhat.com/browse/CNV-75832

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered
- [ ] AI Contributions: The PR abides by the [KubeVirt AI Contribution Policy](https://github.com/kubevirt/community/blob/main/ai-contribution-policy.md).

### Release note

```release-note
Fix VirtualMachineStuckOnNode and VMCannotBeEvicted alerts failing during live migration due to duplicate kubevirt_vmi_info series
```

### CODE REVIEW

**Comment by sradco** [comment]
@machadovilaca @avlitman please review

**Comment by sradco** [comment]
@machadovilaca @avlitman please review this PR

File: pkg/monitoring/rules/alerts/vms.go [hunk 1]
```diff
@@ -153,7 +153,8 @@ var vmsAlerts = []promv1.Rule{
 			"sum by (name, namespace, status, node)((kubevirt_vm_info{status='starting'} == 1 " +
 				"or kubevirt_vm_info{status='stopping'} == 1 or kubevirt_vm_info{status='terminating'} == 1 " +
 				"or (kubevirt_vm_info{status_group='error'} == 1 " +
-				"and on(name, namespace) kubevirt_vmi_info) ) * on(name, namespace) group_left(node) kubevirt_vmi_info)",
+				"and on(name, namespace) kubevirt_vmi_info{phase='running'}) ) " +
+				"* on(name, namespace) group_left(node) topk by(name, namespace) (1, kubevirt_vmi_info{phase='running'}))",
```
Comment (machadovilaca) [inline pkg/monitoring/rules/alerts/vms.go]: what is the rationale behind using only phase running here?

Comment (sradco) [reply pkg/monitoring/rules/alerts/vms.go]: Hi @machadovilaca , This is a node-scoped alert.
we only join against `kubevirt_vmi_info{phase='running'}` because that is the safest phase for meaningful node attribution.
It also avoids duplicate VMI series during transitions/migrations.
For `VirtualMachineStuckOnNode`, phase='running' is the safest filter for an actual node impact

Comment (machadovilaca) [reply pkg/monitoring/rules/alerts/vms.go]: but cant the vmi be in pending or scheduling phases for example?

Comment (sradco) [reply pkg/monitoring/rules/alerts/vms.go]: At pending or scheduling phases the node label is still empty. 
I added `scheduled`, where the pod is actually assigned to the node already, but its a real edge case.

File: hack/prom-rule-ci/prom-rules-tests.yaml [hunk 2]
```diff
@@ -1158,6 +1171,83 @@ tests:
...
+        values: "1x11"
+      # Error with migration overlap: tests inner and + outer group_left
+      - series: 'kubevirt_vm_info{name="vm-error-migration", namespace="test-ns", status="ErrImagePull", status_group="error"}'
+        values: "1x11"
+      - series: 'kubevirt_vmi_info{name="vm-error-migration", namespace="test-ns", phase="running", node="source-node-2"}'
+        values: "1x11"
+      - series: 'kubevirt_vmi_info{name="vm-error-migration", namespace="test-ns", phase="running", node="target-node-2"}'
+        values: "1x11"
+
+    promql_expr_test:
```
Comment (machadovilaca) [inline hack/prom-rule-ci/prom-rules-tests.yaml]: why are you testing alerts using `promql_expr_test` and not the recommended `alert_rule_test` as we usually do
https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#alert_test_case

File: pkg/monitoring/rules/alerts/vms.go [hunk 3]
```diff
@@ -153,7 +153,8 @@ var vmsAlerts = []promv1.Rule{
 			"sum by (name, namespace, status, node)((kubevirt_vm_info{status='starting'} == 1 " +
 				"or kubevirt_vm_info{status='stopping'} == 1 or kubevirt_vm_info{status='terminating'} == 1 " +
 				"or (kubevirt_vm_info{status_group='error'} == 1 " +
-				"and on(name, namespace) kubevirt_vmi_info) ) * on(name, namespace) group_left(node) kubevirt_vmi_info)",
+				"and on(name, namespace) kubevirt_vmi_info{phase=~'scheduled|running'}) ) " +
+				"* on(name, namespace) group_left(node) topk by(name, namespace) (1, kubevirt_vmi_info{phase=~'scheduled|running'}))",
```
Comment (machadovilaca) [inline pkg/monitoring/rules/alerts/vms.go:160]: what happens during migrations?

the node label might be from either the old or the new vmi, no?

that can be misleading if anyone uses the node label for investigating the issue

**Review by machadovilaca — COMMENTED** [review]
I think a release note is needed for this change

PR description
```
After this PR:
Fix VirtualMachineStuckOnNode and VMCannotBeEvicted alerts that were impacted by this.
```

you mentioned VMCannotBeEvicted, either we should also fix here or let's remove it

File: pkg/monitoring/rules/alerts/vms.go (hunk 3 above)
Comment (sradco) [reply pkg/monitoring/rules/alerts/vms.go:160]: This fix is meant to prevent the many-to-many crash and the node is "best effort" in that edge case. The description already says "This may indicate issues with the VM lifecycle on the target node" which is reasonable.
The alternative would be to not include node in the alert at all during migration, but that reduces the alert's value for all non-migration cases.

File: hack/prom-rule-ci/prom-rules-tests.yaml (hunk 2 above)
Comment (sradco) [reply hack/prom-rule-ci/prom-rules-tests.yaml]: updated

**Comment by machadovilaca** [comment]
please check the linter failure

**Review by machadovilaca — APPROVED** [review]
